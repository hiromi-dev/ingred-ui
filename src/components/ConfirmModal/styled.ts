import styled, { keyframes } from "styled-components";
import { Size, Radius } from "../../styles";
import Card from "../Card";

export const Container = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: ${({ theme }) => theme.depth.modal};
`;

const fadeIn = keyframes`
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
`;
export const ModalBackground = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  /* TODO: color */
  background-color: #001326cb;
  animation: ${fadeIn} 0.4s;
`;

const slideIn = keyframes`
  0% {
    transform: translate(-50%, calc(-50% + 8px));
  }
  100% {
    transform: translate(-50%, -50%);
  }
`;
export const ModalContainer = styled(Card)<{ fullSize: boolean }>`
  position: absolute;
  top: 50%;
  left: 50%;
  min-width: 400px; /* 削除モーダルなどコンテンツが少ない場合の最小単位を入れておく(400は仮) */
  width: ${({ fullSize }) => (fullSize ? "90vw" : "auto")};
  height: ${({ fullSize }) => (fullSize ? "90vh" : "auto")};
  transform: translate(-50%, -50%);
  border-radius: ${Radius.MEDIUM};
  background-color: ${({ theme }) => theme.palette.background.default};
  animation: ${slideIn} 0.4s;
`;

export const ScrollContainer = styled.div<{ overflowYScroll: boolean }>`
  max-height: 80vh;
  padding-bottom: ${({ theme }) =>
    theme.spacing * 2 * 2 +
    40}px; /* ModalFooterの高さ(padding上下 + Button size="medium"の高さ) */
  overflow-y: ${({ overflowYScroll }) =>
    overflowYScroll ? "scroll" : "visible"};
`;

export const ModalHeader = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: ${({ theme }) => theme.spacing * 3}px;
  border-bottom: ${Size.Border.Small} solid
    ${({ theme }) => theme.palette.gray.light};
`;

export const TitleContainer = styled.div`
  max-width: 70%;
`;

export const ModalFooter = styled.div`
  position: fixed;
  left: 0;
  bottom: 0;
  display: flex;
  justify-content: flex-end;
  align-items: center;
  width: 100%;
  padding: ${({ theme }) => theme.spacing * 2}px
    ${({ theme }) => theme.spacing * 4}px;
  background-color: ${({ theme }) => theme.palette.gray.light};
  border-radius: 0 0 ${Radius.MEDIUM} ${Radius.MEDIUM};
`;

export const IconContainer = styled.div`
  cursor: pointer;
`;

export const LoadingContainer = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(255, 255, 255, 0.5);
  display: flex;
  justify-content: center;
  align-items: center;
`;
