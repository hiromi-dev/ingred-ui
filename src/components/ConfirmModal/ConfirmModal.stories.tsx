import * as React from "react";
import styled from "styled-components";
import ConfirmModal from ".";
import Spacer from "../Spacer";
import DataTable from "../DataTable";
import { data } from "../DataTable/data";

const Container = styled.div`
  padding: ${({ theme }) => theme.spacing * 3}px;
  background-color: ${({ theme }) => theme.palette.background.dark};
`;

export default {
  title: "ConfirmModal",
  parameters: {
    component: ConfirmModal
  }
};

export const Overview = () => (
  <Container>
    <ConfirmModal title="タイトル">コンテンツ</ConfirmModal>
  </Container>
);

export const WithDanger = () => (
  <Container>
    <ConfirmModal title="タイトル" buttonColor="danger" confirmText="削除する">
      コンテンツ
    </ConfirmModal>
  </Container>
);

export const WithFullSize = () => (
  <Container>
    <Spacer pt={2} />
    <ConfirmModal title="タイトル" fullSize={true}>
      コンテンツ
    </ConfirmModal>
  </Container>
);

export const WithLoading = () => (
  <Container>
    <Spacer pt={2} />
    <ConfirmModal title="タイトル" loading={true}>
      コンテンツ
    </ConfirmModal>
  </Container>
);

export const WithOverflowYScroll = () => (
  <Container>
    <Spacer pt={2} />
    <ConfirmModal
      title="タイトル"
      overflowYScroll={true}
      disableHorizontalPadding={true}
    >
      <DataTable
        data={data}
        columns={[
          {
            name: "ID",
            selector: data => data.id,
            sortable: true
          },
          {
            name: "名前",
            selector: data => data.date,
            sortable: true
          }
        ]}
      />
    </ConfirmModal>
  </Container>
);
