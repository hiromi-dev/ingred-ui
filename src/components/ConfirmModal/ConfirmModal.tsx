import * as React from "react";
import * as Styled from "./styled";
import Typography from "../Typography";
import Icon from "../Icon";
import Spacer from "../Spacer";
import Flex from "../Flex";
import Button from "../Button";
import { ButtonColor } from "../Button/Button";
import Spinner from "../Spinner";
import { useTheme } from "../../themes";

export type Props = {
  title: string;
  confirmText?: string;
  onClose?: () => void;
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void;
  buttonColor?: Exclude<ButtonColor, "cancel">;
  disabled?: boolean;
  loading?: boolean;
  overflowYScroll?: boolean;
  fullSize?: boolean;
  disableHorizontalPadding?: boolean;

  // TypeScriptで型エラーが出るので一旦これでしのぐ
  children?: React.ReactNode;
};

const ConfirmModal: React.FunctionComponent<Props> = ({
  title,
  confirmText = "はい",
  children,
  onClose,
  onSubmit,
  buttonColor = "primary",
  disabled,
  loading,
  fullSize = false,
  overflowYScroll = true,
  disableHorizontalPadding = false
}) => {
  const theme = useTheme();
  return (
    <Styled.Container>
      <Styled.ModalBackground />
      <Styled.ModalContainer fullSize={fullSize}>
        <form onSubmit={onSubmit}>
          <Styled.ScrollContainer overflowYScroll={overflowYScroll}>
            <Styled.ModalHeader>
              <Styled.TitleContainer>
                <Typography weight="bold" size="xxxl">
                  {title}
                </Typography>
              </Styled.TitleContainer>
              <Styled.IconContainer onClick={onClose}>
                <Icon name="close" size="lg" color={theme.palette.black} />
              </Styled.IconContainer>
            </Styled.ModalHeader>

            {disableHorizontalPadding ? (
              children
            ) : (
              <Spacer px={3} pb={3}>
                {children}
              </Spacer>
            )}
          </Styled.ScrollContainer>
          <Styled.ModalFooter>
            <Flex display="flex" alignItems="center">
              <Spacer pr={2}>
                <Button type="button" color="cancel" onClick={onClose}>
                  キャンセル
                </Button>
              </Spacer>
              <Button type="submit" color={buttonColor} disabled={disabled}>
                {confirmText}
              </Button>
            </Flex>
          </Styled.ModalFooter>
        </form>
        {loading && (
          <Styled.LoadingContainer>
            <Spinner />
          </Styled.LoadingContainer>
        )}
      </Styled.ModalContainer>
    </Styled.Container>
  );
};

export default ConfirmModal;
