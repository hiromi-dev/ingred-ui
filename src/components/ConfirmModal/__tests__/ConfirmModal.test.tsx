import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import ConfirmModal from "..";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("ConfirmModal component testing", () => {
  afterEach(cleanup);

  test("ConfirmModal", () => {
    const { asFragment } = renderWithThemeProvider(
      <ConfirmModal title="タイトル" />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
