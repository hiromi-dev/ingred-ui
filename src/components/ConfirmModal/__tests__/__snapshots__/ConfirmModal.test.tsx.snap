// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmModal component testing ConfirmModal 1`] = `
<DocumentFragment>
  <div
    class="sc-AxhCb ivzUGN"
  >
    <div
      class="sc-AxhUy hrDucc"
    />
    <div
      class="sc-AxiKw sc-AxgMl feJJIS"
    >
      <form>
        <div
          class="sc-AxheI Pewgk"
        >
          <div
            class="sc-Axmtr kYXfQl"
          >
            <div
              class="sc-AxmLO eECFKd"
            >
              <p
                class="sc-fzplWN bLugHJ"
                color="#041C33"
                font-size="20px"
              >
                タイトル
              </p>
            </div>
            <div
              class="sc-fzoLsD bpXaOq"
            >
              <div
                class="sc-fznyAO haJfle"
                size="24"
              >
                <svg
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <rect
                    fill="#959FA9"
                    height="23.141"
                    transform="translate(18.364 2) rotate(45)"
                    width="5.142"
                  />
                  <rect
                    fill="#959FA9"
                    height="23.141"
                    transform="translate(2 5.636) rotate(-45)"
                    width="5.143"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            class="sc-AxjAm gPrILy"
          />
        </div>
        <div
          class="sc-fzozJi hvdymy"
        >
          <div
            class="sc-AxirZ beOrDm"
            display="flex"
          >
            <div
              class="sc-AxjAm gJyBXz"
            >
              <button
                class="sc-fznKkj sc-fznZeY bZnhNi"
                font-size="14px"
                font-weight="normal"
                height="42px"
                type="button"
              >
                キャンセル
              </button>
            </div>
            <button
              class="sc-fznKkj sc-fznZeY hUEhlJ"
              font-size="14px"
              font-weight="bold"
              height="42px"
              type="submit"
            >
              はい
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
